gpflow/conditionals/util.py

```diff
@@ -294,66 +352,87 @@ def fully_correlated_conditional_repeat(
-    fmean = tf.linalg.matmul(f, A, transpose_a=True)  # [R, M]  *  [M, K]  ->  [R, K]
-    fmean = tf.reshape(fmean, (R, N, K))  # [R, N, K]
+    fmean = tf.linalg.matmul(f, A, transpose_a=True)  # [R, M]  *  [M, P]  ->  [R, P]
+    fmean = tf.reshape(fmean, (R, N, P))  # [R, N, P]
 
     if q_sqrt is not None:
         Lf = tf.linalg.band_part(q_sqrt, -1, 0)  # [R, M, M]
         if q_sqrt.shape.ndims == 3:
-            A_tiled = tf.tile(A[None, :, :], tf.stack([R, 1, 1]))  # [R, M, K]
-            LTA = tf.linalg.matmul(Lf, A_tiled, transpose_a=True)  # [R, M, K]
+            A_tiled = tf.tile(A[None, :, :], tf.stack([R, 1, 1]))  # [R, M, P]
+            LTA = tf.linalg.matmul(Lf, A_tiled, transpose_a=True)  # [R, M, P]
         elif q_sqrt.shape.ndims == 2:  # pragma: no cover
             raise NotImplementedError("Does not support diagonal q_sqrt yet...")
         else:  # pragma: no cover
             raise ValueError(f"Bad dimension for q_sqrt: {q_sqrt.shape.ndims}")
 
         if full_cov and full_output_cov:
-            addvar = tf.linalg.matmul(LTA, LTA, transpose_a=True)  # [R, K, K]
-            fvar = fvar[None, :, :, :, :] + tf.reshape(addvar, (R, N, K, N, K))
+            addvar = tf.linalg.matmul(LTA, LTA, transpose_a=True)  # [R, P, P]
+            fvar = fvar[None, :, :, :, :] + tf.reshape(addvar, (R, N, P, N, P))
         elif full_cov and not full_output_cov:
-            LTAr = tf.transpose(tf.reshape(LTA, [R, M, N, K]), [0, 3, 1, 2])  # [R, K, M, N]
-            addvar = tf.linalg.matmul(LTAr, LTAr, transpose_a=True)  # [R, K, N, N]
-            fvar = fvar[None, ...] + addvar  # [R, K, N, N]
+            LTAr = tf.transpose(tf.reshape(LTA, [R, M, N, P]), [0, 3, 1, 2])  # [R, P, M, N]
+            addvar = tf.linalg.matmul(LTAr, LTAr, transpose_a=True)  # [R, P, N, N]
+            fvar = fvar[None, ...] + addvar  # [R, P, N, N]
         elif not full_cov and full_output_cov:
-            LTAr = tf.transpose(tf.reshape(LTA, (R, M, N, K)), [0, 2, 3, 1])  # [R, N, K, M]
-            fvar = fvar[None, ...] + tf.linalg.matmul(LTAr, LTAr, transpose_b=True)  # [R, N, K, K]
+            LTAr = tf.transpose(tf.reshape(LTA, (R, M, N, P)), [0, 2, 3, 1])  # [R, N, P, M]
+            fvar = fvar[None, ...] + tf.linalg.matmul(LTAr, LTAr, transpose_b=True)  # [R, N, P, P]
         elif not full_cov and not full_output_cov:
-            addvar = tf.reshape(tf.reduce_sum(tf.square(LTA), axis=1), (R, N, K))  # [R, N, K]
-            fvar = fvar[None, ...] + addvar  # [R, N, K]
+            addvar = tf.reshape(tf.reduce_sum(tf.square(LTA), axis=1), (R, N, P))  # [R, N, P]
+            fvar = fvar[None, ...] + addvar  # [R, N, P]
     else:
         fvar = tf.broadcast_to(fvar[None], tf.shape(fmean))
+
+    shape_constraints.extend(
+        [(Knn, intended_cov_shape), (fmean, ["R", "N", "P"]), (fvar, ["R"] + intended_cov_shape),]
+    )
+    tf.debugging.assert_shapes(shape_constraints, message="fully_correlated_conditional_repeat()")
```

Comment:
no need to change it, cos I'd rather the PR's in, but I would have thought the stack trace would say what function it was raised in

Reply:
you'd hope so , but in tensorflow function land forget about stacktraces ...